### DISCUSSION TITLE
Negative Nodal Areas

### DISCUSSION THREAD

**Discussion by maxposchmann** [idaholab/moose#27149] [Q&A Modules: Solid mechanics]
I've been using the `Contact` action, which generates an auxvariable `nodal_area` calculated by the `NodalArea` userobject. My system is 3D, my mesh uses HEX20 elements, and I was hoping to have everything second order.

Looking at my contact pressures, I noticed that they alternated large positive and negative values from one node to the next. Going a bit further, I noticed the same thing was happening with `nodal_area` (though with signs reversed). I tried going back to a simple generated mesh with just `nodal_area` calculated and the problem persists for second order elements in 3D.

![image](https://github.com/idaholab/moose/assets/44649744/9b28cacc-1184-4532-9064-525412009914)

I'm not actually sure this is incorrect, but it feels wrong in my bones. Negative areas aren't physical. However, the sum of the `nodal_area` values is correct (checked with postprocessor). On the other hand, there are many (infinite?) possible weightings of how the sum is done that would give the correct total area and not have any negative `nodal_area` values.

I tried messing around with quadrature options and the results are the same (obviously the integral over a constant converges quickly).

Any thoughts? Should I try to stop thinking about this? Would an alternative weighting of the area sum be more realistic?

**Comment by maxposchmann** [comment]
Here's an input for testing.
```
[Mesh]
  [./square]
    type = GeneratedMeshGenerator
    nx = 2
    ny = 2
    nz = 2
    dim = 3
    elem_type = HEX20
  [../]
[]

[Variables]
  [./dummy]
  [../]
[]

[AuxVariables]
  [./nodal_area]
    order = SECOND
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./dummy]
    type = Diffusion
    variable = dummy
  [../]
[]

[UserObjects]
  [./nodal_area]
    type = NodalArea
    variable = nodal_area
    boundary = 1
    execute_on = 'initial timestep_end'
  [../]
[]

[BCs]
  [./dummy]
    type = DirichletBC
    variable = dummy
    boundary = 1
    value = 100
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -ksp_gmres_restart'
  petsc_options_value = 'jacobi   101'

  line_search = 'none'

  nl_abs_tol = 1e-11
  nl_rel_tol = 1e-10

  l_max_its = 20
  [Quadrature]
    order = SECOND
    side_order = SECOND
  []
[]

[Postprocessors]
  [face_area]
    type = NodalSum
    variable = nodal_area
    boundary = 1
  []
[]

[Outputs]
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
@recuero @dewenyushu @gambka

**Comment by maxposchmann** [comment]
I don't know if this answers my questions or not, but I tried messing around with the `nodal_area` sum in the userobject and find that the product `nodal_area * contact_pressure` comes out the same. Is `contact_pressure` just for show?

**Reply by maxposchmann** [reply]
Continuing the conversation with my most esteemed colleague, I've noticed that the determination for contact release is:
`if (-contact_pressure >= _tension_release)`
which means that getting the `contact_pressure` (and therefore the `nodal_area`) right actually does matter.

**Comment by dewenyushu** [comment]
I am not sure what is happening for the nodal area values. 

But, for reducing/eliminating unphysical oscillatory contact pressure values - did you try mortar contact? i.e., have `formation = mortar` in your contact action block

**Reply by maxposchmann** [reply]
Thanks for the suggestion. I actually did try mortar at an earlier point, not in response to this particular issue. I found it very slow relative to a simple penalty/glued formulation so was hoping to get away from it. Maybe I made mistakes that made it slower than it ought to have been, but the real case is also just a large-ish 3D model.

**Reply by maxposchmann** [reply]
Also I'd add that I am not convinced there is actually any problem with oscillations in the contact **forces**. The bigger issue is with release/delamination which is derived directly from the pressures as I noted above.

**Reply by gambka** [reply]
When using Mortar have you removed `[Quadrature]` block from the Executioner?  Also, how did you determine the `order` and `side_order` used when playing with the Quadrature options in the node on face formulation? Have to tried increasing these? I haven't tried it on your problem but I've seen higher order quadrature help with oscillations in the past, particularly in the side_order.

```
  [Quadrature]
    order = THIRD
    side_order = FIFTH
  []
```

Perhaps @bwspenc may have some ideas here.

**Reply by maxposchmann** [reply]
I haven't tested mortar thoroughly at all, I can go back and try that in the real problem.

As I mentioned, the `nodal_area` calculation is the same for all `order` and `side_order` values I've tried. This should, and does, converge very quickly. I do not believe there is an issue with contact force oscillations, these are determined directly from `penetration` and don't reference the `nodal_area`.

I think the release issue still remain no matter what is changed in the input and am therefore leaning towards opening an issue now that I have a better handle on what problems the negative `nodal_area` causes.
